src/orion/client/runner.py

```diff
@@ -0,0 +1,323 @@
+        """Gather the results from each worker asynchronously"""
+        results = []
+        try:
+            results = self.client.executor.async_get(
+                self.futures, timeout=self.gather_timeout
+            )
+        except (KeyboardInterrupt, InvalidResult):
+            self.release_all()
+            raise
+
+        to_be_raised = None
+        log.debug(f"Gathered new results {len(results)}")
+
+        # register the results
+        for result in results:
+            self.free_worker += 1
+            trial = self.pending_trials.pop(result.future)
+
+            if isinstance(result, AsyncResult):
+                # NB: observe release the trial already
+                self.client.observe(trial, result.value)
+                self.trials += 1
+
+            if isinstance(result, AsyncException):
+                if (
+                    isinstance(result.exception, ExecutionError)
+                    and result.exception.return_code == self.interrupt_signal_code
+                ):
+                    to_be_raised = KeyboardInterrupt()
+                    self.client.release(trial, status="interrupted")
+                    continue
+
+                if isinstance(result.exception, InvalidResult):
+                    # stop the optimization process if we received `InvalidResult`
+                    # as all the experiments are assumed to be returning those
+                    to_be_raised = result.exception
+
+                # Regular exception, might be caused by the choosen hyperparameters
+                # themselves rather than the code in particular (like Out of Memory error
+                # for big batch sizes)
+                exception = result.exception
+                self.worker_broken_trials += 1
+                self.client.release(trial, status="broken")
+
+                if self.on_error is None or self.on_error(
+                    self, trial, exception, self.worker_broken_trials
+                ):
+                    log.error(result.traceback)
+
+                else:
+                    log.error(str(exception))
+                    log.debug(result.traceback)
+
+                # if we receive too many broken trials, it might indicate the user script
+                # is broken, stop the experiment and let the user investigate
+                if self.worker_broken_trials >= self.max_broken:
+                    raise BrokenExperiment("Worker has reached broken trials threshold")
+
+        if to_be_raised is not None:
+            log.debug("Runner was interrupted")
+            self.release_all()
+            raise to_be_raised
+
+    def release_all(self):
+        """Release all the trials that were reserved by this runner"""
+        # Sanity check
+        for _, trial in self.pending_trials.items():
+            try:
+                self.client.release(trial)
+            except AlreadyReleased:
+                pass
+
+    def _suggest_trials(self, count):
```

Comment:
Should test here behavior between `count` and `self.pool_size`.

Reply:
`self.pool_size` is actually a dead attribute, it is never used

Reply:
It should be used, see comment [here](https://github.com/Epistimio/orion/pull/684/files/94238e7496426a1d7b5ecf18b2a955d82985adbc#r762077645)

Reply:
Ah, so you used it here: https://github.com/Epistimio/orion/pull/684/files#r770818626

Reply:
should be used now